subprojects/launcher/src/main/java/org/gradle/launcher/cli/CommandLineActionFactory.java

```diff
@@ -125,33 +126,49 @@ public WelcomeMessageAction(BuildLayoutParameters buildLayoutParameters) {
 
         @Override
         public void execute(PrintStream out) {
-            GradleVersion currentVersion = GradleVersion.current();
-            File markerFile = getMarkerFile(currentVersion);
+            if (isWelcomeMessageEnabled()) {
+                GradleVersion currentVersion = GradleVersion.current();
+                File markerFile = getMarkerFile(currentVersion);
 
-            if (!markerFile.exists()) {
-                out.println();
-                out.print("Welcome to Gradle " + currentVersion.getVersion() + "!");
+                if (!markerFile.exists()) {
+                    out.println();
+                    out.print("Welcome to Gradle " + currentVersion.getVersion() + "!");
 
-                String featureList = readReleaseFeatures();
+                    String featureList = readReleaseFeatures();
 
-                if (featureList != null) {
-                    out.println();
-                    out.println();
-                    out.print("Here are the highlights of this release:");
-                    out.println();
-                    out.print(featureList);
-                }
+                    if (featureList != null) {
+                        out.println();
+                        out.println();
+                        out.print("Here are the highlights of this release:");
+                        out.println();
+                        out.print(featureList);
+                    }
+
+                    if (!currentVersion.isSnapshot()) {
+                        out.println();
+                        out.print("For more details see https://gradle.org/releases/#" + currentVersion.getVersion());
+                        out.println();
+                    }
 
-                if (!currentVersion.isSnapshot()) {
-                    out.println();
-                    out.print("For more details see https://gradle.org/releases/#" + currentVersion.getVersion());
                     out.println();
+
+                    writeMarkerFile(markerFile);
                 }
+            }
+        }
 
-                out.println();
+        /**
+         * The system property is set for the purpose of internal testing.
+         * In user environment the system property will never be available.
```

Comment:
"In user environment" -> "In user environments,"

Reply:
done